File: backend/internal/dto/equipment.go
package dto

import "time"

// Organization DTOs

type BaseRequest struct {
	Code string `json:"code" binding:"required"`
	Name string `json:"name" binding:"required"`
}

type BaseResponse struct {
	ID        uint      `json:"id"`
	Code      string    `json:"code"`
	Name      string    `json:"name"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type FactoryRequest struct {
	BaseID uint   `json:"base_id" binding:"required"`
	Code   string `json:"code" binding:"required"`
	Name   string `json:"name" binding:"required"`
}

type FactoryResponse struct {
	ID        uint      `json:"id"`
	BaseID    uint      `json:"base_id"`
	BaseName  string    `json:"base_name,omitempty"`
	Code      string    `json:"code"`
	Name      string    `json:"name"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type WorkshopRequest struct {
	FactoryID uint   `json:"factory_id" binding:"required"`
	Code      string `json:"code" binding:"required"`
	Name      string `json:"name" binding:"required"`
}

type WorkshopResponse struct {
	ID         uint      `json:"id"`
	FactoryID  uint      `json:"factory_id"`
	FactoryName string   `json:"factory_name,omitempty"`
	Code       string    `json:"code"`
	Name       string    `json:"name"`
	CreatedAt  time.Time `json:"created_at"`
	UpdatedAt  time.Time `json:"updated_at"`
}

// Equipment Type DTOs

type EquipmentTypeRequest struct {
	Name string `json:"name" binding:"required"`
	Category string `json:"category"`
	InspectionTemplateID *uint `json:"inspection_template_id"`
}

type EquipmentTypeResponse struct {
	ID          uint   `json:"id"`
	Name        string `json:"name"`
	Category    string `json:"category"`
	TemplateID  *uint  `json:"inspection_template_id"`
	CreatedAt   time.Time `json:"created_at"`
}

// Equipment DTOs

type EquipmentRequest struct {
	Code                   string     `json:"code" binding:"required"`
	Name                   string     `json:"name" binding:"required"`
	TypeID                 uint       `json:"type_id" binding:"required"`
	WorkshopID             uint       `json:"workshop_id" binding:"required"`
	Spec                   string     `json:"spec"`
	PurchaseDate           *time.Time `json:"purchase_date"`
	Status                 string     `json:"status"`
	DedicatedMaintenanceID *uint      `json:"dedicated_maintenance_id"`
}

type EquipmentResponse struct {
	ID                      uint       `json:"id"`
	Code                    string     `json:"code"`
	Name                    string     `json:"name"`
	TypeID                  uint       `json:"type_id"`
	TypeName                string     `json:"type_name,omitempty"`
	WorkshopID              uint       `json:"workshop_id"`
	WorkshopName            string     `json:"workshop_name,omitempty"`
	FactoryID               uint       `json:"factory_id,omitempty"`
	FactoryName             string     `json:"factory_name,omitempty"`
	QRCode                  string     `json:"qr_code"`
	Spec                    string     `json:"spec"`
	PurchaseDate            *time.Time `json:"purchase_date"`
	Status                  string     `json:"status"`
	DedicatedMaintenanceID  *uint      `json:"dedicated_maintenance_id"`
	DedicatedMaintenanceName string    `json:"dedicated_maintenance_name,omitempty"`
	CreatedAt               time.Time  `json:"created_at"`
	UpdatedAt               time.Time  `json:"updated_at"`
}

type EquipmentListResponse struct {
	Total int64                `json:"total"`
	Items []EquipmentResponse  `json:"items"`
}

type EquipmentQuery struct {
	Page       int    `form:"page" binding:"min=1"`
	PageSize   int    `form:"page_size" binding:"min=1,max=100"`
	Code       string `form:"code"`
	Name       string `form:"name"`
	TypeID     *uint  `form:"type_id"`
	FactoryID  *uint  `form:"factory_id"`
	WorkshopID *uint  `form:"workshop_id"`
	Status     string `form:"status"`
}

type QRCodeRequest struct {
	EquipmentID []uint `json:"equipment_ids" binding:"required"`
}

type QRCodeResponse struct {
	EquipmentID uint   `json:"equipment_id"`
	Code        string `json:"code"`
	Name        string `json:"name"`
	QRCodeURL   string `json:"qr_code_url"`
	QRCodeData  string `json:"qr_code_data"` // Base64 encoded image
}

// Batch Import DTOs

type EquipmentImportItem struct {
	Code        string `json:"code" binding:"required"`
	Name        string `json:"name" binding:"required"`
	TypeCode    string `json:"type_code" binding:"required"`
	FactoryCode string `json:"factory_code" binding:"required"`
	WorkshopCode string `json:"workshop_code" binding:"required"`
	Spec        string `json:"spec"`
}

type BatchImportRequest struct {
	Items []EquipmentImportItem `json:"items" binding:"required"`
}

type BatchImportResponse struct {
	SuccessCount int      `json:"success_count"`
	FailedCount  int      `json:"failed_count"`
	Errors       []string `json:"errors"`
}

// Organization Tree Response

type OrganizationTree struct {
	ID       uint                `json:"id"`
	Code     string              `json:"code"`
	Name     string              `json:"name"`
	Type     string              `json:"type"` // base, factory, workshop
	Children []OrganizationTree  `json:"children"`
}

// Equipment Statistics

type EquipmentStatistics struct {
	Total      int64 `json:"total"`
	Running    int64 `json:"running"`
	Stopped    int64 `json:"stopped"`
	Maintenance int64 `json:"maintenance"`
	Scrapped   int64 `json:"scrapped"`
}

type EquipmentTypeStatistics struct {
	TypeName  string `json:"type_name"`
	Count     int64  `json:"count"`
	Running   int64  `json:"running"`
	Maintenance int64 `json:"maintenance"`
}

File: frontend/src/api/equipment.ts
import request from './request'

// Types
export interface Base {
  id: number
  code: string
  name: string
  created_at: string
  updated_at: string
}

export interface Factory {
  id: number
  base_id: number
  base_name?: string
  code: string
  name: string
  created_at: string
  updated_at: string
}

export interface Workshop {
  id: number
  factory_id: number
  factory_name?: string
  code: string
  name: string
  created_at: string
  updated_at: string
}

export interface EquipmentType {
  id: number
  name: string
  category?: string
  inspection_template_id?: number
  created_at: string
}

export interface Equipment {
  id: number
  code: string
  name: string
  type_id: number
  type_name?: string
  workshop_id: number
  workshop_name?: string
  factory_id?: number
  factory_name?: string
  qr_code: string
  spec?: string
  purchase_date?: string
  status: 'running' | 'stopped' | 'maintenance' | 'scrapped'
  dedicated_maintenance_id?: number
  dedicated_maintenance_name?: string
  created_at: string
  updated_at: string
}

export interface EquipmentListResponse {
  total: number
  items: Equipment[]
}

export interface EquipmentStatistics {
  total: number
  running: number
  stopped: number
  maintenance: number
  scrapped: number
}

// Organization APIs
export const orgApi = {
  // Bases
  getBases: () => request.get<Base[]>('/organization/bases'),
  createBase: (data: { code: string; name: string }) => request.post<Base>('/organization/bases', data),
  updateBase: (id: number, data: { code: string; name: string }) => request.put(`/organization/bases/${id}`, data),
  deleteBase: (id: number) => request.delete(`/organization/bases/${id}`),

  // Factories
  getFactories: (baseId?: number) =>
    request.get<Factory[]>('/organization/factories', { params: baseId ? { base_id: baseId } : {} }),
  createFactory: (data: { base_id: number; code: string; name: string }) =>
    request.post<Factory>('/organization/factories', data),
  updateFactory: (id: number, data: { base_id: number; code: string; name: string }) =>
    request.put(`/organization/factories/${id}`, data),
  deleteFactory: (id: number) => request.delete(`/organization/factories/${id}`),

  // Workshops
  getWorkshops: (factoryId?: number) =>
    request.get<Workshop[]>('/organization/workshops', {
      params: factoryId ? { factory_id: factoryId } : {},
    }),
  createWorkshop: (data: { factory_id: number; code: string; name: string }) =>
    request.post<Workshop>('/organization/workshops', data),
  updateWorkshop: (id: number, data: { factory_id: number; code: string; name: string }) =>
    request.put(`/organization/workshops/${id}`, data),
  deleteWorkshop: (id: number) => request.delete(`/organization/workshops/${id}`),
}

// Equipment Type APIs
export const equipmentTypeApi = {
  getTypes: () => request.get<EquipmentType[]>('/equipment/types'),
  createType: (data: { name: string; category?: string }) =>
    request.post<EquipmentType>('/equipment/types', data),
  updateType: (id: number, data: { name: string; category?: string }) =>
    request.put(`/equipment/types/${id}`, data),
  deleteType: (id: number) => request.delete(`/equipment/types/${id}`),
}

// Equipment APIs
export const equipmentApi = {
  getList: (params: {
    page?: number
    page_size?: number
    code?: string
    name?: string
    type_id?: number
    factory_id?: number
    workshop_id?: number
    status?: string
  }) => request.get<EquipmentListResponse>('/equipment', { params }),

  getById: (id: number) => request.get<Equipment>(`/equipment/${id}`),

  getByQRCode: (code: string) => request.get<Equipment>(`/equipment/qr/${code}`),

  create: (data: {
    code: string
    name: string
    type_id: number
    workshop_id: number
    spec?: string
    purchase_date?: string
    status?: string
    dedicated_maintenance_id?: number
  }) => request.post<Equipment>('/equipment', data),

  update: (
    id: number,
    data: {
      code: string
      name: string
      type_id: number
      workshop_id: number
      spec?: string
      purchase_date?: string
      status?: string
      dedicated_maintenance_id?: number
    }
  ) => request.put<Equipment>(`/equipment/${id}`, data),

  delete: (id: number) => request.delete(`/equipment/${id}`),

  getStatistics: () => request.get<EquipmentStatistics>('/equipment/statistics'),
  getTypes: equipmentTypeApi.getTypes,
}
